spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java

```diff
@@ -172,18 +173,34 @@ private String jobDesc() {
     Column nameEqual = actualFileName.equalTo(validFileName);
     Column actualContains = actualFileDF.col("file_path").contains(validFileDF.col("file_path"));
     Column joinCond = nameEqual.and(actualContains);
-    List<String> orphanFiles = actualFileDF.join(validFileDF, joinCond, "leftanti")
-        .as(Encoders.STRING())
-        .collectAsList();
-
-    Tasks.foreach(orphanFiles)
-        .noRetry()
-        .executeWith(deleteExecutorService)
-        .suppressFailureWhenFinished()
-        .onFailure((file, exc) -> LOG.warn("Failed to delete file: {}", file, exc))
-        .run(deleteFunc::accept);
-
-    return new BaseDeleteOrphanFilesActionResult(orphanFiles);
+    Dataset<String> orphanDF = actualFileDF.join(validFileDF, joinCond, "leftanti")
+            .as(Encoders.STRING());
+    return deleteFiles(orphanDF);
+  }
+
+  private BaseDeleteOrphanFilesActionResult deleteFiles(Dataset<String> orphanDF) {
+    boolean streamResults = PropertyUtil.propertyAsBoolean(options(), STREAM_RESULTS, false);
+    Iterable<String> itr;
+    if (streamResults) {
+      // cache df to avoid recomputing
+      orphanDF.cache();
+      itr = () -> orphanDF.toLocalIterator();
+    } else {
+      itr = orphanDF.collectAsList();
+    }
+    Tasks.foreach(itr)
+            .noRetry()
+            .executeWith(deleteExecutorService)
+            .suppressFailureWhenFinished()
+            .onFailure((file, exc) -> LOG.warn("Failed to delete file: {}", file, exc))
+            .run(file -> {
+              deleteFunc.accept(file);
+            });
+    if (streamResults) {
```

Comment:
This may end up being very expensive since the code in the Procedure use this iterable a bunch of times. It also ends up putting everything in memory when it build the output rows ... InternalRow[] :/


One thing I think we should still be thinking about is moving all the delete operations to the executors. I know we had a little chat about this on Iceberg Slack and I think it's worth thinking about.

Reply:
I think so. For some tables in our cluster, we have to 'moving all the delete operations to the executors' and use streaming results to avoid oom and accelerate the job.

Reply:
@RussellSpitzer Or we can just return the size of deleted file list instead of the list of files when call the procedure

Reply:
The RewriteDataFiles and RewriteManifestFiles Action are the stream-result option also required ？@RussellSpitzer

Reply:
@microbearz changing that api for that may be have to wait for a version change ... Your fix would change the situation for the action version. Although we probably can't cache like we are currently doing, I think this would leave the uncache for garbage collection of the object on the driver (if it gets uncached at all) which could be a very long time from the completion of the action depending on the use case.